tiatoolbox/tools/pyramid.py

```diff
@@ -430,3 +433,251 @@ def tile_path(self, level: int, x: int, y: int) -> Path:
+        baseline_y = (y * self.tile_size * scale) - (self.overlap * scale)
+        output_size = [self.output_tile_size] * 2
+        coord = [baseline_x, baseline_y]
+        if level < self.sub_tile_level_count:
+            output_size = self.output_tile_size // 2 ** (
+                self.sub_tile_level_count - level
+            )
+            output_size = np.repeat(output_size, 2).astype(int)
+            thumb = self.get_thumb_tile()
+            thumb.thumbnail(output_size)
+            return thumb
+        slide_dimensions = np.array(self.info.slide_dimensions)
+        if all(slide_dimensions < [baseline_x, baseline_y]):
+            raise IndexError
+
+        rgb = np.zeros((*output_size, 4), dtype=np.uint8)
+        bounds = locsize2bounds(coord, [self.tile_size * scale] * 2)
+        bound_geom = Polygon.from_bounds(*bounds)
+        rgb = self.render_annotations(rgb, bound_geom, scale, coord)
+
+        return Image.fromarray(rgb)
+
+    def render_by_type(
+        self,
+        rgb: np.ndarray,
+        ann: Annotation,
+        top_left: Tuple[float, float],
+        scale: int,
+        poly_as_box: bool = False,
+    ):
+        """Render annotation appropriately to its geometry type.
+
+        Args:
+            rgb (np.ndarray):
+                The image to render the annotation on.
+            ann (Annotation):
+                The annotation to render.
+            top_left (Tuple[int, int]):
+                The top left coordinate of the tile.
+            scale (int):
+                The scale at which we are rendering the tile.
+            poly_as_box (bool):
+                Whether to render polygons as boxes.
+
+        """
+        r = self.renderer
+        geom_type = ann.geometry.geom_type
+        if geom_type == "Point":
+            r.render_pt(rgb, ann, top_left, scale)
+        elif geom_type == "Polygon":
+            if poly_as_box:
+                r.render_rect(rgb, ann, top_left, scale)
+            else:
+                r.render_poly(rgb, ann, top_left, scale)
+        elif "Line" in geom_type:
+            r.render_line(rgb, ann, top_left, scale)
+        else:
+            warnings.warn("Unknown geometry")
+
+    def render_annotations(
+        self,
+        rgb: np.ndarray,
+        bound_geom: Polygon,
+        scale: int,
+        top_left: Tuple[float, float],
+    ):
+        """Get annotations as bbox or geometry according to zoom level,
+        and render them, decimating large collections of small annotations
+        if appropriate.
+
+        Args:
+            rgb (np.ndarray):
+                The image to render the annotation on.
+            bound_geom (Polygon):
+                A polygon representing the bounding box of the tile.
+            scale (int):
+                The scale at which we are rendering the tile.
+            top_left (Tuple[int, int]):
+                The top left coordinate of the tile.
```

Comment:
I'm a bit unclear as to why both `bound_geom` and `top_left` are given here. Should one not be derived from the other? Is this intended to be able to render an arbitrary region? If so then why is the top_left needed? Is it intended to render a specific tile? If so then why not take in `x, y, level` as with the get tile function?

Reply:
It is redundant. I can change that.

Reply:
Have removed the redundant top_left, renamed ann to annotation and renamed rgb to tile (the function is get_tile so I thought it made most sense to have the thing that is being passed around and eventually returned be called tile too)